### DISCUSSION TITLE
Extracting dynamic meta?

### DISCUSSION THREAD

**Discussion by joshuaiz** [manovotny/extract-mdx-metadata#21] [Q&A]
First, thank you for this package - was exactly what I needed.

This is more of a question than an issue but I was trying to use this in a `.mdx` file that had dynamic data from imports like so:

```js
import { joshuaiz } from '@/authors'
import Waveform from '../../../components/Waveform'
import image from './card.jpg'

export const meta = {
  title: "Encapsulating Markup and Styles in Shopify with CodeKit's Hooks and SCSS",
  description: "Keeping your styles with the theme files that use them using CodeKit's Hooks.",
  date: '2020-06-20T18:05:31Z',
  authors: [joshuaiz], // this generates an error
  image, // so does this
}

// rest of content
```
The dynamic values for `authors` and `image` above generate an error with `extract-mdx-metadata`.

As a workaround I changed the values in the `meta` object to strings which works but I was wondering if it is possible to have dynamically generated data extracted?

My guess is no because there is no discreet runtime step to grab the dynamic data...is that correct?

Otherwise, if it *is* possible, it would be a great feature to have.

**Comment by manovotny** [comment]
Intriguing... I don't know if it's possible or not, but there's only one way to find out! 😅 Let me give it a shot and see what I can do. Thanks for raising the question!

**Reply by manovotny** [reply]
@joshuaiz I am here to tell you, I think ***it is possible***! 🎉 

A couple of updates...

* The solution has me caught in this whole ESM special kinda hell. I am still working my way out of it, but it's taking me a bit of time to come up to speed. It's a nasty cycle of ESM at the future and the path forward, but tooling is lagging support making the path forward difficult. I'll get it...eventually.
* The `image` part of this is interesting and I'd like your take on it... What would you expect it to return? A path to the image or a data url / uri? Choose your answer carefully. 😉

**Reply by joshuaiz** [reply]
@manovotny this is great news. Yeah ESM is the wild west right now but the right way to go.

As for the `image` return: good question. Let me go through and check what it returns now and get back to you. I don't want to choose the wrong thing :)

**Reply by joshuaiz** [reply]
Just to update this, currently I'm importing the path as a string like so:

```js
import image from './card.jpg'
```

That path then gets transpiled by Vercel in Next.js to something like this:

```js
https://xxxxxxxxxx.com/_next/static/media/card.13c1d2ad2f27ff397e207a3c9eaadd79.jpg
```

So as long as that behavior can be maintained, I think importing as a dynamic string (src) is the way.

**Reply by manovotny** [reply]
@joshuaiz here's where I'm at and what I'm thinking...

While Next.js is a huge driver of MDX, MDX does exist apart from Next.js. This project is currently built in a way where it can process any MDX, not just Next.js MDX. I am very hesitate to couple those two together, so I don't think I'll be implementing anything that is Next.js specific. Additionally, given that the assets are hashed at Next.js build / run time (ie. `card.13c1d2ad2f27ff397e207a3c9eaadd79.jpg`), and there's no good way (as far as I know) to hook into the build / run time to get access to the randomly generated hash, I don't know how I'd even support keeping the hashes in sync with one another.

I will say that this is also an interesting way to do this in Next.js. I guess I’ve always just placed images in the `public` / static folder and referenced them as strings instead of importing them like code. I am not saying what you’re doing is wrong or incorrect. Just haven’t personally done it like that before and maybe that’s a recommendation that could work for you in this scenario.

That said, I see two viable paths forward and could expose one or both via options. I’d like to hear your thoughts.

#### Convert images to [data uri / url](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URIs)

The resulting uri / url are just strings, meaning it would be usable in many contexts, including using it as a `src` to an `img`, passable / storable anywhere, etc. 

However, data uris / urls are large, and depending on how that data is used, could result in bloating bundle sizes. They’re also not catchable, at least in the traditional sense (cached as a part of the bundle, but not as a standalone asset).

#### Using public path

When it finds an import it would prepend a specific public path. 

For example, if you passed the public path of `https://example.com/images` and it found an import like `import image from './card.jpg’`, it would produce `https://example.com/images/card.jpg`.

**Reply by joshuaiz** [reply]
@manovotny I think I misspoke. I didn't mean for you to support Next.js' hashing but more to support the image path as string rather than having to encode the image as data uri. Next.js will do what it does with that string so you don't need to get involved in that.

And yes, we'd want to make this as platform-agnostic as possible.

Keeping the imports as a string should make things easier in the long run. Sorry for any confusion.

**Reply by manovotny** [reply]
@joshuaiz I've released [v3.0.0-beta.1](https://github.com/manovotny/extract-mdx-metadata/releases/tag/v3.0.0-beta.1) for testing. 

I've skipped any image support, for now, as it increase complexity dramatically. 

I would love it if you could give it a shot and let me know how it goes. I fully expect that I didn't get it right on the first try. 😉 

You can view the new readme [here](https://github.com/manovotny/extract-mdx-metadata/blob/fabd015714f5693730da416ea8fac1dfb43fe8b1/readme.md).

**Reply by manovotny** [reply]
Have you had a chance to try this @joshuaiz? @drewkiimon, @irenajur, and @ncarn2, I also see that you 👍 the original question. Could any of you try the beta referenced above and provide feedback?

**Reply by etnbrd** [reply]
Hey 👋 
I got in this situation as well, and might have some clarifications.

## Dynamic meta

The current implementation of `extract-mdx-metadata`, if I understand correctly, parses the AST of the mdx file, grab the `meta` variable definition, and `eval` it, to get the evaluated definition of this variable: https://github.com/manovotny/extract-mdx-metadata/blob/main/index.js#L21

So, basically, dynamic meta is already _partially_ supported by `extract-mdx-metadata`, but not to our satisfaction:
```typescript
const outside = 'outside' // outside the variable declaration node AST
export const meta = {
  works: 'static string',
  alsoWorks: `selfcontained string: ${Math.random()}`,
  wontWork: outside // Error: outside is not defined (or something)
}
```

How MDx usually works is through a compiler: MDX in, JSX out, like [`@mdx-js/mdx`](https://www.npmjs.com/package/@mdx-js/mdx) or [`xdm`](https://www.npmjs.com/package/xdm).
What we are trying to achieve here is the evaluation of the resulting JSX, rather than just the compilation.

`xdm` allows to do that with [`evaluate`](https://www.npmjs.com/package/xdm#evaluatefile-options), but with some limitations:
> Typically, import (or export … from) do not work here. They can be compiled to dynamic import() by passing options.useDynamicImport.
Which brings us to the issue with `import`.

## Imported meta

Because MDX is compiled to another JSX file, the resulting JSX file (or the file evaluating the JSX) might not be in the same folder, or within the same environment. So the import paths might be misleading. Basically, the import path in the source depends on the compilation output directory, which should not be dependent at all.

Usually, it's not a problem because the compilation is done through a builder, such as webpack, or esbuild, which uses loaders that are aware of this issue. They might fix the import paths, or they might compile and output the module that is imported to a similar relative path.
But if some MDX turned JSX is evaluated outside of its hierarchy, any import path is probably going to miss.
We might fix the import before the evaluation, or pass some prefix to fix the paths of some dynamic import, but that's some tedious work that is already implemented in the compilers or builders.

@manovotny, v3.0.0-beta.1 seems to use `esbuild` with the `jsx` loader and the `xdm` compiler to compile a specific mdx file, along with all the imported dependencies, bundle it into a single javascript string. Then this string is passed to `requiredFromString` to be evaluated.
It indeed fix the issue of evaluating MDX outside of its hierarchy, but it'll also surely bring some performance issue, as all the components are compiled and bundled, for each MDX file.

If the MDX content is compiled and evaluated somewhere else, it doubles the compilation and evaluation.
Moreover, I guess the options for the content compiler and the meta compiler might need to be synchronized or shared somehow, as some options might apply to both the content and meta.
Which brings us to the issue with custom image loaders (one of the aforementioned options for the compiler).

## Image custom loaders

Next.js, which is (or used to be?) based on webpack, provides some custom loaders for images (and other resources), to allow importing everything from js.
When `import image from './image.jpg'` returns a string, what it actually does behind the scene, is call a custom loader.
When compiling files, webpack (and other compilers) will look for `require` or `import` statement, and resolve the imported files. If these files have certain extensions, they will use specific loaders. The [next.js image custom loader](https://www.npmjs.com/package/next-images) will copy the resolved file to a public location, maybe it will resize it, or inline it if it's small enough, and instead of returning the content of the file, it'll return the public path.

So, if `extract-mdx-metadata` would want to support that, it would need to accept these loaders as parameters.

So, to summarize, `extract-mdx-metadata` will need to bring an MDX compiler (e.g. `@mdx-js/mdx` or `xdm`), a JSX evaluator (e.g. `eval`, `requireFromString`, `import ()` ...), and support custom loaders from the platform within which it is used (e.g. `next.js`, `webpack` ...).
All these will most likely be different from the ones that are used in the project from within `extract-mdx-metadata` is used, so it'd be hard to be really platform agnostic.

Given that the whole MDX file would be compiled and evaluated anyway to extract meta, I guess it would make sense to use the same pipeline for the processing (compilation, evaluation, resolution) of both the meta and the content.

**Reply by manovotny** [reply]
Howdy y'all! 👋 I haven't given up or forgotten about this. There's just been other things and..."life". 😅 

I decided to reach out to @wooorm and get his opinion since so much has changed in the MDX space since I first wrote this in 2019. 

After some back and forth DMs and some [experimentation](https://github.com/manovotny/mdx-metadata-extraction-experiments), I don't believe this package is needed anymore in the near-ish future. Tooling in the space is evolving enough to support this natively. 

The easiest path forward seems to be using [xdm's `evaluate`](https://github.com/wooorm/xdm#evaluatefile-options), which you can see in [this implementation and working example](https://github.com/manovotny/mdx-metadata-extraction-experiments/blob/main/scripts/evaluate.js).

Two things to note:

1. It is using [Node’s ESM loader](https://nodejs.org/dist/latest-v17.x/docs/api/esm.html#loaders) functionality and it is still experimental.
2. It also needs / relies on [a custom JSX loader](https://github.com/mdx-js/mdx/blob/main/script/jsx-loader.js) that is currently unpublished and copied / pasted in React / markdown / MDX land.

But all-in-all, this kind of solution would consumers the ability to customize a solution that meets their explicit need rather than `extract-mdx-metadata` trying to guess the need of many use cases.

Thoughts?

**Reply by etnbrd** [reply]
I arrived to the same conclusion, and have been using a similar solution :)
